server/src/main/java/org/opensearch/action/support/replication/ReplicationOperation.java

```diff
@@ -226,13 +232,164 @@ private void performOnReplicas(
+            final ShardRouting primaryRouting
+        );
+    }
+
+    /**
+     * This implementation of {@link ReplicationProxy} fans out the replication request to current shard routing if
+     * it is not the primary and has replication mode as {@link ReplicationMode#LOGICAL_REPLICATION}.
+     *
+     * @opensearch.internal
+     */
+    private class FanoutReplicationProxy extends ReplicationProxy {
+
+        private FanoutReplicationProxy() {
+
+        }
+
+        @Override
+        ReplicationMode determineReplicationMode(ReplicationModeAwareShardRouting shardRouting, ShardRouting primaryRouting) {
+            return shardRouting.getShardRouting().isSameAllocation(primaryRouting) == false
+                ? ReplicationMode.LOGICAL_REPLICATION
+                : ReplicationMode.NONE;
+        }
+    }
+
+    /**
+     * This implementation of {@link ReplicationProxy} fans out the replication request to current shard routing basis
+     * the shard routing's replication mode and replication override policy.
+     *
+     * @opensearch.internal
+     */
+    private class ReplicationModeAwareOverrideProxy extends ReplicationProxy {
+
+        private final ReplicationOverridePolicy overridePolicy;
+
+        private ReplicationModeAwareOverrideProxy(ReplicationOverridePolicy overridePolicy) {
+            assert Objects.nonNull(overridePolicy);
+            this.overridePolicy = overridePolicy;
+        }
+
+        @Override
+        ReplicationMode determineReplicationMode(ReplicationModeAwareShardRouting shardRouting, ShardRouting primaryRouting) {
+            ShardRouting currentRouting = shardRouting.getShardRouting();
+
+            // If the current routing is the primary, then it does not need to be replicated
+            if (currentRouting.isSameAllocation(primaryRouting)) {
+                return ReplicationMode.NONE;
+            }
+
+            // If the current routing's replication mode is not NONE, then we return the original replication mode.
+            if (shardRouting.getReplicationMode() != ReplicationMode.NONE) {
+                return shardRouting.getReplicationMode();
+            }
+
+            // If the current routing's replication mode is none, then we check for override and return overridden mode.
+            if (overridePolicy.allowOverride) {
+                return overridePolicy.overriddenMode;
+            }
+
+            // At the end, return NONE.
+            return ReplicationMode.NONE;
+        }
+    }
+
+    /**
+     * Defines the replication override policy which individual {@link TransportReplicationAction} can implement.
+     *
+     * @opensearch.internal
+     */
+    public static class ReplicationOverridePolicy {
+
+        private final boolean allowOverride;
+
+        private final ReplicationMode overriddenMode;
+
+        public ReplicationOverridePolicy(boolean allowOverride, ReplicationMode overriddenMode) {
```

Comment:
I don't think you need both allowOverride and overriddenMode: if `overriddenMode` is `false`, the `overriddenMode` has no use.

Reply:
Ah, yes. Earlier I was returning a non-null static variable in the `getReplicationOverridePolicy()` method. Made change to return null. 
I will remove the `overriddenMode` field. However, will keep it as a class for extensibility and ease of adding fields in future.